### PR TITLE
`defer_measurements` raises errors for unsupported measurements

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Measurements that cover all wires can give users unexpected outputs when using mid-circuit measurements. So, `defer_measurements` now explicitly removes support for such measurements by raising errors.

**Description of the Change:**
* Raise an error if `qml.state()` is used.
* Raise an error if `qml.probs()`, `qml.sample()`, or `qml.counts()` is used if wires or an observable isn't specified.
* Add warning to docs

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
#4363

### CODE REVIEW

**Review by albi3ro — APPROVED** [review]
I know there's a lot of debate over this, but I'm fine with this decision and its better than people being confused about what their answers actually mean.

**Review by timmysilv — APPROVED** [review]

**Comment by mudit2812** [comment]
> I know there's a lot of debate over this, but I'm fine with this decision and its better than people being confused about what their answers actually mean.

For reference, @trbromley and I had a discussion offline and the decision was to go in this direction in the end.

**Comment by Alex-Preciado** [comment]
@mudit2812 Thank you for validating with @trbromley and for posting the update here.

File: pennylane/transforms/defer_measurements.py [hunk 1]
```diff
@@ -23,7 +23,62 @@
...
+    if ops_cv or obs_cv:
+        raise ValueError("Continuous variable operations and observables are not supported.")
+
+    for mp in tape.measurements:
+        if isinstance(mp, (CountsMP, ProbabilityMP, SampleMP)) and not (
+            mp.obs or mp._wires or mp.mv
+        ):
+            raise ValueError(
+                f"Cannot use {mp.__class__.__name__} as a measurement without specifying wires "
+                "when using qml.defer_measurements."
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:47]: Can this warning come up if users are just using mid-circuit measurements without explicitly requesting to defer? If so, I'd have:
```suggestion
                "when using qml.defer_measurements. Deferred measurements can occur "
                "automatically when using mid-circuit measurements on a device that does not "
                "support them."
```

File: pennylane/transforms/defer_measurements.py [hunk 2]
```diff
@@ -23,7 +23,62 @@
...
+            )
+
+        if mp.__class__.__name__ == "StateMP":
+            raise ValueError(
+                "Cannot use StateMP as a measurement when using qml.defer_measurements."
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:52]:
```suggestion
                "Cannot use StateMP as a measurement when using qml.defer_measurements. "
                "Deferred measurements can occur automatically when using mid-circuit "
                "measurements on a device that does not support them."
```

File: pennylane/transforms/defer_measurements.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:47]: Thanks Tom!

**Comment by trbromley** [comment]
Looking back at this PR, we probably should have given a _recommendation_, i.e., if you're seeing an error because of `qml.state()`, try using `qml.density_matrix()` instead.

**Pull request merged**